components/patient-data/rest/src/main/java/org/phenotips/data/rest/internal/DefaultPatientsFetchResourceImpl.java

```diff
@@ -0,0 +1,222 @@
+     * Retrieves and collects patient entities that correspond to the provided external ID.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param eid an external patient ID, as string
+     */
+    private void collectPatientsFromEid(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final Object eid)
+    {
+        try {
+            final Query q = qm.createQuery("where doc.object(PhenoTips.PatientClass).external_id = :eid", Query.XWQL);
+            q.bindValue("eid", eid);
+            final List<Object> patientIds = q.execute();
+            addIds(patientsBuilder, patientIds);
+        } catch (final QueryException ex) {
+            logger.warn("Failed to retrieve patient with external id [{}]: {}", eid, ex.getMessage());
+        }
+    }
+
+    /**
+     * Retrieves patient entities given a list of internal patient IDs.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param ids a list of patient ids, as strings
+     */
+    private void addIds(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final List<Object> ids)
+    {
+        for (final Object id : ids) {
+            if (StringUtils.isNotBlank((String) id)) {
+                addPatientFromId(patientsBuilder, id);
+            }
+        }
+    }
+
+    /**
+     * Given the patient's internal ID, retrieves the patient entity, if it exists and if the user has view rights, and
+     * adds it to the set of patient entities.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param id an internal patient ID
+     */
+    private void addPatientFromId(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final Object id)
+    {
+        try {
+            // Try to get the patient entity.
+            final PrimaryEntity patient = this.repository.get((String) id);
+            // Get the current user.
+            final User currentUser = this.users.getCurrentUser();
+            // If user has view rights and patient with the provided ID exists, add patient to patient set.
+            if (patient != null && this.access.hasAccess(Right.VIEW, currentUser == null ? null
+                : currentUser.getProfileDocument(), patient.getDocument())) {
+                patientsBuilder.add(patient);
+            }
+        } catch (final IllegalArgumentException ex) {
+            logger.warn("Failed to retrieve patient with ID [{}]: {}", id, ex.getMessage());
+        }
+    }
+
+    /**
+     * A custom object mapper to facilitate serializing a list of {@link Patient} objects.
+     *
+     * @return an object mapper that can serialize {@link Patient} objects
+     */
+    private static ObjectMapper getCustomObjectMapper()
+    {
+        final ObjectMapper objectMapper = new ObjectMapper();
+        SimpleModule m = new SimpleModule(SERIALIZER_LABEL, new Version(1, 0, 0, "", SERIALIZER_LABEL, "json"));
```

Comment:
Still not right:

    public Version(int major, int minor, int patchLevel, String snapshotInfo, String groupId, String artifactId)

Should be something like:

    new Version(1, 0, 0, "", "org.phenotips", "PatientReferenceSerializer"));